### PR TITLE
Assume effects

### LINKED ISSUE #769: compat for @assume_effects
introduced in https://github.com/JuliaLang/julia/pull/43852

### PR DESCRIPTION
**Pull request by Tokazama**
Fixes #769 

This allows packages with a 1.6 compat to replace `Base.@pure` with `Base.@assume_effects :total` similar to what's been done in base. All other effects are ignored

### CODE REVIEW

**Comment by Keno** [comment]
`@pure` is stronger than `@assume_effects`. I don't think this is a good idea.

**Comment by Tokazama** [comment]
> `@pure` is stronger than `@assume_effects`. I don't think this is a good idea.

Sorry, I thought this was how all uses of `Base.@pure` were managed in Base. What's the appropriate way to transition code that uses `Base.@pure`?

**Comment by Keno** [comment]
It should be generally safe to replace calls to `@pure`, by `@assume_effects :total`, but not the other way around. The issue is that with this, you'll write `@assume_effects :total`, but get `@pure` semantics on older versions of julia, which is not the same.

**Comment by Keno** [comment]
More generally, though, packages should revisit every use of `@pure` and select the minimum set of effect assumptions that they need (preferably none).

**Comment by Tokazama** [comment]
I remove the mapping to `@pure` so there aren't any added assumptions on 1.6.
I guess there could be an extended help section for `Base.@pure` detailing how to make the transition but a quick search on [JuliaHub](https://juliahub.com/ui/Search?q=%40pure&type=code) reveals that most packages that still use it are maintained by one particular individual. If I recall correctly, previous attempts to dissuade their use of `@pure` were met with quite a bit of resistance.

File: src/Compat.jl [hunk 1]
```diff
@@ -51,6 +51,15 @@ end
     end
 end
 
+# https://github.com/JuliaLang/julia/pull/43852
+if isdefined(Base, Symbol("@assume_effects"))
```
Comment (martinholters) [inline src/Compat.jl:55]:
```suggestion
if VERSION < v"1.8.0-DEV.1484"
```
Makes it easier in the future to figure out whether we can remove this when we bump the minimum Julia version.

File: src/Compat.jl [hunk 2]
```diff
@@ -51,6 +51,15 @@ end
     end
 end
 
+# https://github.com/JuliaLang/julia/pull/43852
+if VERSION < v"1.8.0-DEV.1484"
```
Comment (martinholters) [inline src/Compat.jl:55]:
```suggestion
@static if VERSION < v"1.8.0-DEV.1484"
```
should make this work even if `@assume_effects` is undefined.

Comment (martinholters) [reply src/Compat.jl]: No, scratch that, it's just that the two branches of the `if` need to be swapped, right? Before `v"1.8.0-DEV.1484"`, we want to define our own macro, after we use the one from `Base`. Not the other way round.

File: src/Compat.jl (hunk 1 above)
Comment (martinholters) [reply src/Compat.jl]: Whoops, that reverses the logic though, sorry about that.

**Review by martinholters — APPROVED** [review]
LGTM, but we should get #784 in first and then test on top of that to verify it's working as intended on latest Julia.

That said, I wonder how helpful this is given it cannot be used as a replacement for `@pure` without regressions on older Julia. Still better than nothing, though.

@Keno any more thoughts on this?

**Comment by martinholters** [comment]
Can you rebase on (or merge) current master to (hopefully) fix the test failures on nightly?

**Pull request merged**